File: src/main/java/myratpackexamples/webpoll/RetrievePollHandler.java
package myratpackexamples.webpoll;

import com.google.inject.Inject;
import io.netty.handler.codec.http.HttpHeaderNames;
import io.netty.handler.codec.http.HttpResponseStatus;
import lombok.AllArgsConstructor;
import lombok.Value;
import ratpack.handling.Context;
import ratpack.handling.Handler;
import ratpack.jackson.Jackson;

@Value
@AllArgsConstructor(onConstructor=@__(@Inject))
public class RetrievePollHandler implements Handler {
    PollRepository pollRepository;

    @Override
    public void handle(Context context) throws Exception {
        String pollId = context.getPathTokens().get("poll");

        pollRepository.retrievePoll(pollId)
                .then(validation -> validation
                        .toEither()
                        .peek(poll -> createSuccessResponse(context, poll))
                        .peekLeft(error -> createFailureResponse(context, error))
                );
    }

    private static void createSuccessResponse(Context context, Poll poll) {
        context.getResponse().getHeaders().add(HttpHeaderNames.LOCATION, "poll/" + poll.getId());
        context.getResponse().status(HttpResponseStatus.OK.code());
        context.render(Jackson.json(poll));
    }

    private static void createFailureResponse(Context context, String error) {
        if (error.equals("RatpackMongoClient.ExactlyOneElementExpected()")) {
            context.getResponse().status(HttpResponseStatus.NOT_FOUND.code());
        } else if (error.equals("Non valid id!")) {
            context.getResponse().status(HttpResponseStatus.BAD_REQUEST.code());
        } else {
            context.getResponse().status(HttpResponseStatus.INTERNAL_SERVER_ERROR.code());
        }
        context.getResponse().send("");
    }
}

File: src/main/java/myratpackexamples/webpoll/RetrievePollHandler.kt
package myratpackexamples.webpoll

import com.google.inject.Inject
import io.netty.handler.codec.http.HttpHeaderNames
import io.netty.handler.codec.http.HttpResponseStatus
import io.netty.handler.codec.http.HttpResponseStatus.*
import myratpackexamples.webpoll.FindOneError.ExactlyOneElementExpected
import myratpackexamples.webpoll.FindOneError.InvalidIdString
import ratpack.handling.Context
import ratpack.handling.Handler
import ratpack.jackson.Jackson

class RetrievePollHandler @Inject constructor(val pollRepository: PollRepository) : Handler {

    override fun handle(context: Context) {
        val pollId = context.pathTokens["poll"]

        pollRepository.retrievePoll(pollId)
                .then {
                    it.toEither()
                            .peek(context::createSuccessResponse)
                            .peekLeft(context::createFailureResponse)
                }
    }
}

private fun Context.createSuccessResponse(poll: PollResponse.Poll) {
    response.headers.add(HttpHeaderNames.LOCATION, "poll/" + poll.id)
    response.status(HttpResponseStatus.OK.code())
    render(Jackson.json(poll))
}

private fun Context.createFailureResponse(error: FindOneError) {
    response.status(mapErrorToResponseCode(error).code())
    response.send("")
}

private fun mapErrorToResponseCode(error: FindOneError): HttpResponseStatus {
    return when (error) {
        is ExactlyOneElementExpected -> NOT_FOUND
        is InvalidIdString -> BAD_REQUEST
        else -> INTERNAL_SERVER_ERROR
    }
}
